accepted/wasmtime-one-dot-oh.md

```diff
@@ -0,0 +1,558 @@
+  and new feature development.
+* Released with Wasmtime - these components are all released on the same cadence
+  as the rest of the Wasmtime project (more on this cadence later).
+* Production quality - this component is well vetted and reviewed by multiple
+  developers. Landing changes requires code review and extensive test suites to
+  pass.
+
+At the time of this RFC, the only component in the Wasmtime ecosystem which
+meets these criteria is the `wasmtime` Rust crate itself. This tier is intended
+to be quite strict in terms of requirements and nontrivial to reach, and the
+other components of Wasmtime today do not yet currently meet all of these
+requirements and consequently will fall into lower tiers.
+
+#### Tier 2 - API unstable, production quality
+
+This second tier of support differentiates itself from tier 1 components not
+requiring as much API stability. The same level of production quality for code
+and implementation, however, is expected of tier 2 components. The checklist for
+this tier is:
+
+* API unstable - strict API stability is not required at this time. Components
+  can freely have major changes to their API without much warning. This is
+  intended to be an incubation area for production quality implementations which
+  haven't yet necessarily had enough experience to determine the best API yet.
+  Additionally RFCs are not required for changes in this tier.
+* Well-maintained - same as tier 1.
+* Released with Wasmtime - same as tier 1.
+* Production quality - same as tier 1.
+
+Wasmtime components that fall into this category are the C API of Wasmtime and
+the `wasmtime-wasi` crate. In both of these cases the API stability isn't quite
+there yet, but both are well-reviewed and production quality.
+
+Note that no language bindings are yet to the Tier 2 level yet. This is
+intentional and discussed in the next section.
+
+### Tier 3 - API unstable, not guaranteed production-ready
+
+This is the lowest of the tiers of supported APIs for Wasmtime and is intended
+to be a sort of "catch-all" for everything that doesn't fall into the above
+tiers. This tier does not come with many guarantees associated with it, and is
+intended to be a breeding ground and call-for-help for components to reach tier
+2 or higher status.
+
+Components in tier 3 may still meet various requirements of tier 1 and 2 while
+still being classified as tier 3. For example components here may not have code
+review but could still be of a high-enough quality to use in production.
+Similarly they might be released at the same cadence of Wasmtime itself but
+lacking in API stability.
+
+Today this tier includes:
+
+* `wasmtime-py` - Python bindings for Wasmtime
+* `wasmtime-go` - Go bindings for Wasmtime
+* `wasmtime-dotnet` - .NET bindings for Wasmtime
+* `wasmtime-cpp` - C++ bindings for Wasmtime
+* `witx-bindgen` - Canonical API bindings generator for Interface Types
```

Comment:
Since you've included `witx-bindgen` here, is this expected to be used by all Bytecode Alliance projects or only by Wasmtime (with sub-crates) and its language bindings?

Reply:
Great question! We do expect `witx-bindgen` to be used by other projects (both in and outside the BA), so in some sense it wouldn't need to have a release schedule aligned with Wasmtime's, and there's nothing tying `witx-bindgen` to Wasmtime in some fundamental way.

At the same time, I think there are practical reasons for doing so, at least for the time being. As of now, simply as a matter of fact we're making heavy use of Wasmtime in the development process of witx-bindgen. And increasingly that'll be the case the other way around as well. Aligning the release processes means we can rely on available feature sets in these development processes.

Should over time other uses of `witx-bindgen` become dominant, it'd make sense to reevaluate this alignment.

Reply:
Right, at the same time, for someone using `witx-bindgen` only to generate guest bindings, seeing "random" major version bumps without any significant changes in `witx-bindgen` might be a weird experience.

Reply:
That's true, yes, and to some degree more so than for the Wasmtime embeddings.

Our processes in developing `witx-bindgen` itself would become increasingly more complicated if we didn't align version numbers like this that I think the benefits outweigh these costs, though. Which seems even more true given that, when using it as a guest bindings generator, one should only need to update `witx-bindgen` to pull in significant changes. And even then, the major version upgrade should be trivial—except if `witx-bindgen` has significant API changes which would necessitate a major version bump anyway.

Reply:
Currently `witx-bindgen` has a lot of various components which I suspect is going to end up like the `wasm-tools` repo where each component is separately managed and versioned. I expect anything wasmtime-related to be tied to wasmtime's own version, but things like the JS code generator or Rust generator for compiled-to-wasm-code probably don't need to be tied to weasmtime's own release process.